examples/deployments/router_standalone_trtllm/worker.py

```diff
@@ -0,0 +1,590 @@
+                "kv_cache_usage": kv_cache_usage,
+            }
+        )
+
+    def close(self):
+        self.socket.close()
+        self.context.term()
+
+
+class KvEventsPublisher:
+    """Publishes KV cache events over ZMQ."""
+
+    def __init__(self, port: int, block_size: int):
+        self.context = zmq.Context()
+        self.socket = self.context.socket(zmq.PUB)
+        self.socket.bind(f"tcp://*:{port}")
+        self.block_size = block_size
+        self.partial_block_hashes: set[int] = set()
+        self.sequence_number = 0
+
+    def publish_stored(
+        self,
+        block_hashes: list[int],
+        token_ids: list[int],
+        parent_hash: int | None,
+        block_mm_infos: list[dict | None] | None,
+    ):
+        """Publish a BlockStored event.
+
+        Args:
+            block_hashes: List of block hashes being stored.
+            token_ids: All token IDs across the blocks.
+            parent_hash: Hash of the parent block (if any).
+            block_mm_infos: Per-block multimodal info list. Each element corresponds
+                to a block and is either the mm_info dict (for blocks containing
+                image tokens) or None (for text-only blocks).
+        """
+        event = {
+            "type": "BlockStored",
+            "block_hashes": [to_unsigned_u64(h) for h in block_hashes],
+            "token_ids": token_ids,
+            "block_size": self.block_size,
+        }
+
+        if parent_hash is not None:
+            event["parent_block_hash"] = to_unsigned_u64(parent_hash)
+
+        if block_mm_infos is not None:
+            event["block_mm_infos"] = block_mm_infos
+
+        self._send([event])
+
+    def publish_removed(self, block_hashes: list[int]):
+        """Publish a BlockRemoved event."""
+        # Filter out partial blocks
+        filtered = []
+        for h in block_hashes:
+            if h in self.partial_block_hashes:
+                self.partial_block_hashes.remove(h)
+            else:
+                filtered.append(to_unsigned_u64(h))
+
+        if filtered:
+            self._send([{"type": "BlockRemoved", "block_hashes": filtered}])
+
+    def _send(self, events: list[dict]):
+        """Send events via ZMQ multipart message."""
+        batch = [time.time(), events, 0]
+        try:
+            payload = msgpack.packb(batch, use_bin_type=True)
+        except Exception as e:
+            logger.error(f"msgpack error: {e}")
+            return
+
+        seq_bytes = self.sequence_number.to_bytes(8, byteorder="big")
+        self.sequence_number += 1
+        self.socket.send_multipart([b"", seq_bytes, payload])
+
+    def close(self):
+        self.socket.close()
+        self.context.term()
+
+
+# -----------------------------------------------------------------------------
+# KV Event Processing Helpers
+# -----------------------------------------------------------------------------
+
+
+def extract_mm_info(
```

Comment:
The function is good for the current implementation. But for future, it's probably better for TRTLLM to provide an API to return mm info to reduce integration gap.